syntax = "proto3";

package streaming;

// Service definition
service StreamingService {
  // Basic operations
  rpc ListarTodosUsuarios(Empty) returns (UsuariosResponse);
  rpc ListarTodasMusicas(Empty) returns (MusicasResponse);
  rpc ListarPlaylistsUsuario(UsuarioRequest) returns (PlaylistsResponse);
  rpc ListarMusicasPlaylist(PlaylistRequest) returns (MusicasResponse);
  rpc ListarPlaylistsComMusica(MusicaRequest) returns (PlaylistsResponse);
  rpc ObterEstatisticas(Empty) returns (EstatisticasResponse);
  
  // CRUD operations for Users
  rpc ObterUsuario(UsuarioRequest) returns (Usuario);
  rpc CriarUsuario(CriarUsuarioRequest) returns (Usuario);
  rpc AtualizarUsuario(AtualizarUsuarioRequest) returns (Usuario);
  rpc DeletarUsuario(UsuarioRequest) returns (BooleanResponse);
  
  // CRUD operations for Music
  rpc ObterMusica(MusicaRequest) returns (Musica);
  rpc CriarMusica(CriarMusicaRequest) returns (Musica);
  rpc AtualizarMusica(AtualizarMusicaRequest) returns (Musica);
  rpc DeletarMusica(MusicaRequest) returns (BooleanResponse);
  
  // CRUD operations for Playlists
  rpc ObterPlaylist(PlaylistRequest) returns (Playlist);
  rpc CriarPlaylist(CriarPlaylistRequest) returns (Playlist);
  rpc AtualizarPlaylist(AtualizarPlaylistRequest) returns (Playlist);
  rpc DeletarPlaylist(PlaylistRequest) returns (BooleanResponse);
  
  // Streaming operation
  rpc StreamMusicas(stream MusicaRequest) returns (stream Musica);
}

// Empty message for requests that don't need parameters
message Empty {}

// Request messages
message UsuarioRequest {
  string id_usuario = 1;
}

message PlaylistRequest {
  string id_playlist = 1;
}

message MusicaRequest {
  string id_musica = 1;
}

// CRUD request messages
message CriarUsuarioRequest {
  string nome = 1;
  int32 idade = 2;
}

message AtualizarUsuarioRequest {
  string id_usuario = 1;
  string nome = 2;
  int32 idade = 3;
}

message CriarMusicaRequest {
  string nome = 1;
  string artista = 2;
  int32 duracao_segundos = 3;
}

message AtualizarMusicaRequest {
  string id_musica = 1;
  string nome = 2;
  string artista = 3;
  int32 duracao_segundos = 4;
}

message CriarPlaylistRequest {
  string nome = 1;
  string id_usuario = 2;
  repeated string musicas = 3;
}

message AtualizarPlaylistRequest {
  string id_playlist = 1;
  string nome = 2;
  repeated string musicas = 3;
}

// Boolean response for delete operations
message BooleanResponse {
  bool success = 1;
  string message = 2;
}

// Response messages
message Usuario {
  string id = 1;
  string nome = 2;
  int32 idade = 3;
}

message Musica {
  string id = 1;
  string nome = 2;
  string artista = 3;
  int32 duracao_segundos = 4;
}

message Playlist {
  string id = 1;
  string nome = 2;
  string id_usuario = 3;
  repeated string musicas = 4;
}

message UsuariosResponse {
  repeated Usuario usuarios = 1;
}

message MusicasResponse {
  repeated Musica musicas = 1;
}

message PlaylistsResponse {
  repeated Playlist playlists = 1;
}

message EstatisticasResponse {
  int32 total_usuarios = 1;
  int32 total_musicas = 2;
  int32 total_playlists = 3;
  float media_musicas_por_playlist = 4;
  string tecnologia = 5;
}